pkg/fieldpath/merge.go

```diff
@@ -0,0 +1,156 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fieldpath
+
+import (
+	"reflect"
+
+	"github.com/imdario/mergo"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	errInvalidMerge = "failed to merge values"
+)
+
+// MergeOptions Specifies merge options on a field path
+type MergeOptions struct { // TODO(aru): add more options that control merging behavior
+	// Specifies that already existing values in a merged map should be preserved
+	KeepMapValues bool `json:"keepMapValues,omitempty"`
+	// Specifies that already existing elements in a merged slice should be preserved
+	AppendSlice bool `json:"appendSlice,omitempty"`
+}
+
+// MergoConfiguration the default behavior is to replace maps and slices
+func (mo *MergeOptions) MergoConfiguration() []func(*mergo.Config) {
+	config := []func(*mergo.Config){mergo.WithOverride}
+	if mo == nil {
+		return config
+	}
+
+	if mo.KeepMapValues {
+		config = config[:0]
+	}
+	if mo.AppendSlice {
+		config = append(config, mergo.WithAppendSlice)
+	}
+	return config
+}
+
+// ToPaved tries to convert a runtime.Object into a *Paved via
+// runtime.DefaultUnstructuredConverter if needed. Returns the paved if
+// the conversion is successful along with whether the
+// runtime.DefaultUnstructuredConverter has been employed during the
+// conversion.
+func ToPaved(o runtime.Object) (*Paved, bool, error) {
+	if u, ok := o.(interface{ UnstructuredContent() map[string]interface{} }); ok {
+		return Pave(u.UnstructuredContent()), false, nil
+	}
+
+	oMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(o)
+	if err != nil {
+		return nil, false, err
+	}
+	return Pave(oMap), true, nil
+}
+
+// PatchFieldValueToObject applies the value to the "to" object at the given
+// path with the given merge options, returning any errors as they occur.
+// If no merge options is supplied, then destination field is replaced
+// with the given value.
+func PatchFieldValueToObject(fieldPath string, value interface{}, to runtime.Object,
+	mergeOptions *MergeOptions) error {
+	paved, copied, err := ToPaved(to)
+	if err != nil {
+		return err
+	}
+	dst, err := paved.GetValue(fieldPath)
+	if IsNotFound(err) || mergeOptions == nil {
+		dst = nil
+	} else if err != nil {
+		return err
+	}
+
+	dst, err = merge(dst, value, mergeOptions)
+	if err != nil {
+		return err
+	}
+
+	if err := paved.SetValue(fieldPath, dst); err != nil {
+		return err
+	}
+
+	if copied {
+		return runtime.DefaultUnstructuredConverter.FromUnstructured(paved.object, to)
+	}
+	return nil
+}
+
+// MergePath merges the value at the given field path of the src object into
+// the dst object.
+func MergePath(fieldPath string, dst, src runtime.Object, mergeOptions *MergeOptions) error {
```

Comment:
I'm not sure how I feel about making this package deal more directly with `runtime.Object` - i.e. taking them as an argument. So far the established pattern has been to convert a `runtime.Object` (or anything else) to a `*Paved`, then operate on that. What do you think about implementing `MergePath` as a method on `*Paved` similar to the current `SetValue`? For example:

```go
func (p *Paved) MergeValue(path string, value interface{}, o ...MergeOption) error
```